### PR TITLE
Allow specifying native-image options & new ./gradlew reflectionConfig task

### LINKED ISSUE #15: Allow for native-image arguments in configuration
For some libraries it's necessary to enable the -H:+JNI flag, to be able to build a proper native image. Currently the Gradle nativeImage task doesn't allow for setting this execution parameter. 

Allowing for other configuration like configuration files, could also help (e.g. to set resolve possible reflection issues with a reflect.json)

### PR DESCRIPTION
**Pull request by iamdanfox**
## Before this PR

This plugin couldn't be used to create native images for anything that involves Jackson.

The `-H:ReflectionConfigurationFiles=/path/to/reflectconfig` is necessary to make Jackson deserialization work, but there was no way to specify any kind of extra options to the `native-image` invocation.

## After this PR

You can now create native-images that use ObjectMappers, `@JsonPropery` annotations etc etc 🎉🎉🎉

People can add arbitrary options, e.g.:

```gradle 
        nativeImage {
            option '-H:Optimize=1'
        }
```

Also, you can run `./gradlew reflectionConfig` to generate the necessary JSON from some given class names:

```json
[
  {
    "name" : "java.lang.Class",
    "allDeclaredConstructors" : true,
    "allPublicConstructors" : true,
    "allDeclaredMethods" : true,
    "allPublicMethods" : true
  }
]
```

fixes https://github.com/palantir/gradle-graal/issues/15

### CODE REVIEW

File: src/test/groovy/com/palantir/gradle/graal/GradleGraalEndToEndSpec.groovy [hunk 1]
```diff
@@ -42,6 +42,10 @@ class GradleGraalEndToEndSpec extends IntegrationSpec {
             outputName 'hello-world'
             graalVersion '1.0.0-rc5'
         }
+        
+        nativeImage {
+            option '-H:Optimize=1'
```
Comment (markelliot) [inline src/test/groovy/com/palantir/gradle/graal/GradleGraalEndToEndSpec.groovy:47]: tests should include understanding what happens when provided options conflict with native supported gradle options

Comment (markelliot) [reply src/test/groovy/com/palantir/gradle/graal/GradleGraalEndToEndSpec.groovy:47]: outputName and mainClass are nativeImage options, not general graal options.

I'd also prefer 'nativeImage' configuration live underneath 'graal', for organization.

File: src/main/java/com/palantir/gradle/graal/NativeImageOptions.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+import org.gradle.api.provider.Provider;
+
+/**
+ * See substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/SubstrateOptions.java in
+ * https://github.com/oracle/graal/.
+ */
+public final class NativeImageOptions implements Serializable {
+    private static final long serialVersionUID = 9307652985066589L;
+
+    // using Object to represent String | Provider<String>
```
Comment (markelliot) [inline src/main/java/com/palantir/gradle/graal/NativeImageOptions.java:35]: prefer a union type

File: src/main/java/com/palantir/gradle/graal/GradleGraalPlugin.java [hunk 3]
```diff
@@ -76,6 +76,17 @@ public final void apply(Project project) {
                     task.dependsOn(downloadGraal);
                 });
 
+        TaskProvider<ReflectionConfigTask> reflectionConfig = project.getTasks().register(
```
Comment (markelliot) [inline src/main/java/com/palantir/gradle/graal/GradleGraalPlugin.java:79]: it'd be more appropriate to split this task into a separate PR from the options addition

Comment (markelliot) [reply src/main/java/com/palantir/gradle/graal/GradleGraalPlugin.java:79]: (more broadly, I'd prefer a different strategy to humans adding files to a list, happy to talk about the specifics offline)

Comment (iamdanfox) [reply src/main/java/com/palantir/gradle/graal/GradleGraalPlugin.java:79]: OK, happy to split that into a separate PR.  Agreed that manually curating that list is not sustainable. The strategy I'm leaning towards to generate this reflectconfig class list is to run a little JavaExec task which uses some reflection to find all classes annotated with @JsonProperty, @JsonSerialize etc.

File: src/test/groovy/com/palantir/gradle/graal/GradleGraalEndToEndSpec.groovy (hunk 1 above)
Comment (iamdanfox) [reply src/test/groovy/com/palantir/gradle/graal/GradleGraalEndToEndSpec.groovy:47]: Not sure I understand your feedback here... what's an example of a native supported gradle option that you'd like to avoid conflicting with?

This PR didn't touch the outputName and mainClass properties, but are you requesting that they now live under the nativeImage block? e.g.:

```gradle
graal {
  graalVersion '1.0.0-rc5'
  nativeImage {
    outputName 'hello-world'
    mainClass 'com.palantir.test.Main'
  }
}
```

Comment (markelliot) [reply src/test/groovy/com/palantir/gradle/graal/GradleGraalEndToEndSpec.groovy:47]: outputName causes us to add `-H:Name`, what happens if I also specify `option '-H:Name=blah'`

File: src/main/java/com/palantir/gradle/graal/GradleGraalPlugin.java (hunk 3 above)
Comment (markelliot) [reply src/main/java/com/palantir/gradle/graal/GradleGraalPlugin.java:79]: I think we'd be better off with an explicit "GraalReflectable" annotation + annotation processor than specifically scanning for Jackson annotations.

Comment (markelliot) [reply src/main/java/com/palantir/gradle/graal/GradleGraalPlugin.java:79]: ...and that shouldn't be part of this repo, or generative of a Java class, but instead something that generates the Graal-format config file.

**Comment by helpermethod** [comment]
Hi,

so if I understand this correctly, is currently not possible to create a native image by using a custom 
ReflectionConfigurationFile?

**Comment by markelliot** [comment]
@helpermethod we ended up adding an `option` keyword to the nativeImage configuration, and you can use that to set the reflection configuration file.

**Comment by helpermethod** [comment]
@markelliot Thanks for the answer. Any chance to have the `option` keyword merged soon? I would love to use your plugin but without the ability to specify options to the `nativeImage` extension I can't use it.

**Comment by uschi2000** [comment]
@helpermethod the `option` functionality should be part of the 0.2.0 release, https://github.com/palantir/gradle-graal/pull/22
